<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head><title>R: Core Utilities for Single-Cell RNA-Seq</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>
<h1> Core Utilities for Single-Cell RNA-Seq
<img class="toplogo" src="../../../doc/html/Rlogo.svg" alt="[R logo]" />
</h1>
<hr/>
<div style="text-align: center;">
<a href="../../../doc/html/packages.html"><img class="arrow" src="../../../doc/html/left.jpg" alt="[Up]" /></a>
<a href="../../../doc/html/index.html"><img class="arrow" src="../../../doc/html/up.jpg" alt="[Top]" /></a>
</div><h2>Documentation for package &lsquo;sccore&rsquo; version 0.1.1</h2>

<ul><li><a href="../DESCRIPTION">DESCRIPTION file</a>.</li>
</ul>

<h2>Help Pages</h2>


<table width="100%">
<tr><td style="width: 25%;"><a href="adjacentVertices.html">adjacentVertices</a></td>
<td>List of adjacent vertices from igraph object</td></tr>
<tr><td style="width: 25%;"><a href="adjacent_vertex_weights.html">adjacent_vertex_weights</a></td>
<td>List of adjacent vertex weights from igraph object</td></tr>
<tr><td style="width: 25%;"><a href="appendSpecificityMetricsToDE.html">appendSpecificityMetricsToDE</a></td>
<td>Append specificity metrics to DE</td></tr>
<tr><td style="width: 25%;"><a href="as_factor.html">as_factor</a></td>
<td>convert character vector into a factor with names "values" and "levels"</td></tr>
<tr><td style="width: 25%;"><a href="cellAnnotations.html">cellAnnotations</a></td>
<td>Conos cell annotations</td></tr>
<tr><td style="width: 25%;"><a href="collapseGraphPaga.html">collapseGraphPaga</a></td>
<td>Collapse graph using PAGA 1.2 algorithm, Wolf et al 2019, Genome Biology (2019) &lt;https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1663-x&gt;</td></tr>
<tr><td style="width: 25%;"><a href="collapseGraphSum.html">collapseGraphSum</a></td>
<td>Collapse Graph By Sum</td></tr>
<tr><td style="width: 25%;"><a href="colSumByFac.html">colSumByFac</a></td>
<td>Calculates factor-stratified sums for each column</td></tr>
<tr><td style="width: 25%;"><a href="conosClusterList.html">conosClusterList</a></td>
<td>Conos clusters list</td></tr>
<tr><td style="width: 25%;"><a href="conosGraph.html">conosGraph</a></td>
<td>Conos graph</td></tr>
<tr><td style="width: 25%;"><a href="dotPlot.html">dotPlot</a></td>
<td>Dot plot adapted from Seurat:::DotPlot, see ?Seurat:::DotPlot for details</td></tr>
<tr><td style="width: 25%;"><a href="embeddingColorsPlot.html">embeddingColorsPlot</a></td>
<td>Set colors for embedding plot. Used primarily in embeddingPlot().</td></tr>
<tr><td style="width: 25%;"><a href="embeddingGroupPlot.html">embeddingGroupPlot</a></td>
<td>Plotting function for cluster labels, names contain cell names. Used primarily in embeddingPlot().</td></tr>
<tr><td style="width: 25%;"><a href="embeddingPlot.html">embeddingPlot</a></td>
<td>Plot embedding with provided labels / colors using ggplot2</td></tr>
<tr><td style="width: 25%;"><a href="embedGraphUmap.html">embedGraphUmap</a></td>
<td>Embed a graph into a UMAP, Uniform Manifold Approximation and Projection for Dimension Reduction, &lt;https://github.com/lmcinnes/umap&gt;, &lt;doi:10.21105/joss.00861&gt;</td></tr>
<tr><td style="width: 25%;"><a href="embedKnnGraph.html">embedKnnGraph</a></td>
<td>Embed a k-nearest neighbor (kNN) graph within a UMAP. Used within embedGraphUmap(). Please see McInnes et al &lt;doi:10.21105/joss.00861&gt; for the UMAP description and implementation.</td></tr>
<tr><td style="width: 25%;"><a href="extendMatrix.html">extendMatrix</a></td>
<td>Extend matrix to include new columns in matrix</td></tr>
<tr><td style="width: 25%;"><a href="fac2col.html">fac2col</a></td>
<td>Utility function to translate a factor into colors</td></tr>
<tr><td style="width: 25%;"><a href="fac2palette.html">fac2palette</a></td>
<td>Encodes logic of how to handle named-vector and functional palettes. Used primarily within embeddingGroupPlot()</td></tr>
<tr><td style="width: 25%;"><a href="getClusterGraph.html">getClusterGraph</a></td>
<td>Collapse vertices belonging to each cluster in a graph</td></tr>
<tr><td style="width: 25%;"><a href="get_nearest_neighbors.html">get_nearest_neighbors</a></td>
<td>Get nearest neighbors method on graph</td></tr>
<tr><td style="width: 25%;"><a href="graphToAdjList.html">graphToAdjList</a></td>
<td>Convert igraph graph into an adjacency list</td></tr>
<tr><td style="width: 25%;"><a href="jsDist.html">jsDist</a></td>
<td>Jensen–Shannon distance metric (i.e. the square root of the Jensen–Shannon divergence) between the columns of a dense matrix m</td></tr>
<tr><td style="width: 25%;"><a href="mergeCountMatrices.html">mergeCountMatrices</a></td>
<td>Merge list of count matrices into a common matrix, entering 0s for the missing entries</td></tr>
<tr><td style="width: 25%;"><a href="multi2dend.html">multi2dend</a></td>
<td>Translate multilevel segmentation into a dendrogram, with the lowest level of the dendrogram listing the cells</td></tr>
<tr><td style="width: 25%;"><a href="plapply.html">plapply</a></td>
<td>Parallel, optionally verbose lapply. See ?parallel::mclapply for more info.</td></tr>
<tr><td style="width: 25%;"><a href="propagateLabels.html">propagateLabels</a></td>
<td>Estimate labeling distribution for each vertex, based on provided labels.</td></tr>
<tr><td style="width: 25%;"><a href="propagateLabelsDiffusion.html">propagateLabelsDiffusion</a></td>
<td>Estimate labeling distribution for each vertex, based on provided labels using a Random Walk on graph</td></tr>
<tr><td style="width: 25%;"><a href="propagateLabelsSolver.html">propagateLabelsSolver</a></td>
<td>Propagate labels using Zhu, Ghahramani, Lafferty (2003) algorithm, "Semi-Supervised Learning Using Gaussian Fields and Harmonic Functions" &lt;http://mlg.eng.cam.ac.uk/zoubin/papers/zgl.pdf&gt;</td></tr>
<tr><td style="width: 25%;"><a href="propagate_labels.html">propagate_labels</a></td>
<td>Label propagation</td></tr>
<tr><td style="width: 25%;"><a href="setMinMax.html">setMinMax</a></td>
<td>Set range for values in object. Changes values outside of range to min or max. Adapted from Seurat::MinMax</td></tr>
<tr><td style="width: 25%;"><a href="smooth_count_matrix.html">smooth_count_matrix</a></td>
<td>Smooth gene expression, used primarily within conos::correctGenes. Used to smooth gene expression values in order to better represent the graph structure. Use diffusion of expression on graph with the equation dv = exp(-a * (v + b))</td></tr>
<tr><td style="width: 25%;"><a href="sn.html">sn</a></td>
<td>Set names equal to values, a stats::setNames wrapper function</td></tr>
<tr><td style="width: 25%;"><a href="splitVectorByNodes.html">splitVectorByNodes</a></td>
<td>splitVectorByNodes</td></tr>
<tr><td style="width: 25%;"><a href="styleEmbeddingPlot.html">styleEmbeddingPlot</a></td>
<td>Set plot.theme, legend, ticks for embedding plot. Used primarily in embeddingPlot().</td></tr>
<tr><td style="width: 25%;"><a href="umapEmbedding.html">umapEmbedding</a></td>
<td>UMAP embedding</td></tr>
<tr><td style="width: 25%;"><a href="val2ggcol.html">val2ggcol</a></td>
<td>Helper function to return a ggplot color gradient for a numeric vector ggplot(aes(color=x, ...), ...) + val2ggcol(x)</td></tr>
</table>
</body></html>
